### ISSUE TITLE
train.py is expecting Arrow files and an entropy model but data preparation creates JSONL files

### ISSUE THREAD

**Issue by RefractAI** [facebookresearch/blt#4]
I am following the steps in the readme.

download_prepare_hf_data.py created a folder of JSONL files, but when running train, I get:

Zero shard_files found corresponding to: fineweb_edu_10bt using preprocess_dir=entropy_preprocess and entropy_model_name=transformer_100m

So therefore:
1. How are the arrow files created from the JSONL files?
2. Where can I find transformer_100m for preprocessing?

**Comment by EntilZha** [comment]
Hi! We're still working on getting all the preprocessing steps updated from the Lingua code to our code, sorry for the delay! I'll post here once its up :).

**Comment by nican2018** [comment]
Hi!
Please any update for this?
Have you been able to solve the problem?

**Comment by EntilZha** [comment]
Sorry for the delay, most of use were away the past 2 weeks with US holidays. I'll see what I can do this week to get this working with the data that lingua pulls.

**Comment by EntilZha** [comment]
Hi, I'm currently working on making this easier to do (between holidays and start of year company things, been busy until now). If you want to jump the gun a bit and not wait longer, here is the rough sketch of what I'm planning to do:

1. Double check the lingua download scripts work as expected + create config file to point to them
2. Add a script that trains the small entropy model from this data, which is a pre-requisite to the next step
3. The preprocessing code (given the JSON file and entropy model checkpoint) is in `bytelatent.preprocess.preprocess_entropies`. I need check if I need to make any changes for this to work with our OSS code (mainly around the entropy checkpoint loading). The rest of the code, namely the schema, will be the same and generates the individual arrow files in the list `self.dataset_files`
4. The actual preprocessing is done by `bytelatent.preprocess.parallel_entropies` which runs `bytelatent.preprocess.preprocess_entropies` in parallel. This should work as is, but I can provide some scripts/configs that make it out of the box work with lingua data. It also hardcodes using slurm to parallelize, if this is an issue, there is an alternate executor in `submitit` that just uses python multiprocessing.

That should enable training from lingua data. I'll make incremental updates/PRs so you can make use of this right away as its done versus waiting for the full thing to be done

**Comment by Unpredictable-12** [comment]
Hi！！！！！I'm very interested in this BLT model and want to train and reproduce it.Now，I am following the steps in the readme.
If I want to conduct a complete training of a BLT model now and ultimately obtain a model file while also validating it, what functions do I need to run, or what is the complete process? It's currently a bit confusing, for example, with scripts like `download_prepare_hf_data.py`, `train.py`, etc.

**Comment by EntilZha** [comment]
Hi @Unpredictable-12, I'm working on making this a bit clearer and fixing some scripts that are needed for a full reproduction. In general, the steps will be:

1. Use the HF based download scripts
2. Train the entropy model using the jsonl files
3. Preprocess the HF data to arrow data, which mainly involves ahead-of-time running the entropy model on the training data and saving entropies (alternatively, can run the entropy model online, but we need to port the code to do this).
4. Run train.py

I pushed changes yesterday that fixes the script to preprocess a given jsonl file to arrow for (3), and I'm working adding a config/instructions for training the entropy model now.

**Comment by Unpredictable-12** [comment]
@EntilZha "Hello, you mentioned earlier about training the entropy model to help with the subsequent training of the BLT model, but I haven't found a fast method or code for training the entropy model. Could you please help explain this? Thank you very much!"

**Comment by Unpredictable-12** [comment]
@EntilZha Hello, author! With your guidance, the training process has been going smoothly.
But，I have already trained the entropy model and am currently running the preprocess_entropies.py script. I have a question: What is the difference between the dataset used for training the entropy model and the dataset used to generate the Arrow files when running the preprocess_entropies.py script? Is it:
The two processes require the same dataset?
The datasets used for the two processes are completely different?
The dataset used for training the BLT model includes the dataset used for training the entropy model?
I'm a bit unclear about it at the moment and would really appreciate your help. Thank you so much!

**Comment by Unpredictable-12** [comment]
@EntilZha Dear author! I have now successfully trained the entropy model and preprocessed the arrow file. However, when running the train.py function of the BLT model, I encountered a problem that has puzzled me for a long time. After all the contents were loaded, an error occurred when running “pred = model(batch_x, patch_lengths=batch_patch_lengths, ngram_ids=ngram_ids)”. Could you please help me solve this? This error has been bothering me for a long time and I haven't been able to solve it. The error message is as follows:Thank you so much!Thank you so much!Thank you so much!！！！
[rank0]: Traceback (most recent call last):
[rank0]:   File "/root/anaconda3/envs/env_blt/lib/python3.10/site-packages/torch/_dynamo/output_graph.py", line 1446, in _call_user_compiler
[rank0]:     compiled_fn = compiler_fn(gm, self.example_inputs())
[rank0]:   File "/root/anaconda3/envs/env_blt/lib/python3.10/site-packages/torch/_dynamo/repro/after_dynamo.py", line 129, in __call__
[rank0]:     compiled_gm = compiler_fn(gm, example_inputs)
[rank0]:   File "/root/anaconda3/envs/env_blt/lib/python3.10/site-packages/torch/__init__.py", line 2235, in __call__
[rank0]:     return compile_fx(model_, inputs_, config_patches=self.config)
[rank0]:   File "/root/anaconda3/envs/env_blt/lib/python3.10/site-packages/torch/_inductor/compile_fx.py", line 1521, in compile_fx
[rank0]:     return aot_autograd(
[rank0]:   File "/root/anaconda3/envs/env_blt/lib/python3.10/site-packages/torch/_dynamo/backends/common.py", line 72, in __call__
[rank0]:     cg = aot_module_simplified(gm, example_inputs, **self.kwargs)
[rank0]:   File "/root/anaconda3/envs/env_blt/lib/python3.10/site-packages/torch/_functorch/aot_autograd.py", line 1071, in aot_module_simplified
[rank0]:     compiled_fn = dispatch_and_compile()
[rank0]:   File "/root/anaconda3/envs/env_blt/lib/python3.10/site-packages/torch/_functorch/aot_autograd.py", line 1056, in dispatch_and_compile
[rank0]:     compiled_fn, _ = create_aot_dispatcher_function(
[rank0]:   File "/root/anaconda3/envs/env_blt/lib/python3.10/site-packages/torch/_functorch/aot_autograd.py", line 522, in create_aot_dispatcher_function
[rank0]:     return _create_aot_dispatcher_function(
[rank0]:   File "/root/anaconda3/envs/env_blt/lib/python3.10/site-packages/torch/_functorch/aot_autograd.py", line 759, in _create_aot_dispatcher_function
[rank0]:     compiled_fn, fw_metadata = compiler_fn(
[rank0]:   File "/root/anaconda3/envs/env_blt/lib/python3.10/site-packages/torch/_functorch/_aot_autograd/jit_compile_runtime_wrappers.py", line 586, in aot_dispatch_autograd
[rank0]:     compiled_fw_func = aot_config.fw_compiler(fw_module, adjusted_flat_args)
[rank0]:   File "/root/anaconda3/envs/env_blt/lib/python3.10/site-packages/torch/_inductor/compile_fx.py", line 1350, in fw_compiler_base
[rank0]:     return _fw_compiler_base(model, example_inputs, is_inference)
[rank0]:   File "/root/anaconda3/envs/env_blt/lib/python3.10/site-packages/torch/_inductor/compile_fx.py", line 1421, in _fw_compiler_base
[rank0]:     return inner_compile(
[rank0]:   File "/root/anaconda3/envs/env_blt/lib/python3.10/site-packages/torch/_inductor/compile_fx.py", line 475, in compile_fx_inner
[rank0]:     return wrap_compiler_debug(_compile_fx_inner, compiler_name="inductor")(
[rank0]:   File "/root/anaconda3/envs/env_blt/lib/python3.10/site-packages/torch/_dynamo/repro/after_aot.py", line 85, in debug_wrapper
[rank0]:     inner_compiled_fn = compiler_fn(gm, example_inputs)
[rank0]:   File "/root/anaconda3/envs/env_blt/lib/python3.10/site-packages/torch/_inductor/compile_fx.py", line 661, in _compile_fx_inner
[rank0]:     compiled_graph = FxGraphCache.load(
[rank0]:   File "/root/anaconda3/envs/env_blt/lib/python3.10/site-packages/torch/_inductor/codecache.py", line 1370, in load
[rank0]:     compiled_graph = compile_fx_fn(
[rank0]:   File "/root/anaconda3/envs/env_blt/lib/python3.10/site-packages/torch/_inductor/compile_fx.py", line 570, in codegen_and_compile
[rank0]:     compiled_graph = fx_codegen_and_compile(gm, example_inputs, **fx_kwargs)
[rank0]:   File "/root/anaconda3/envs/env_blt/lib/python3.10/site-packages/torch/_inductor/compile_fx.py", line 878, in fx_codegen_and_compile
[rank0]:     compiled_fn = graph.compile_to_fn()
[rank0]:   File "/root/anaconda3/envs/env_blt/lib/python3.10/site-packages/torch/_inductor/graph.py", line 1913, in compile_to_fn
[rank0]:     return self.compile_to_module().call
[rank0]:   File "/root/anaconda3/envs/env_blt/lib/python3.10/site-packages/torch/_inductor/graph.py", line 1839, in compile_to_module
[rank0]:     return self._compile_to_module()
[rank0]:   File "/root/anaconda3/envs/env_blt/lib/python3.10/site-packages/torch/_inductor/graph.py", line 1867, in _compile_to_module
[rank0]:     mod = PyCodeCache.load_by_key_path(
[rank0]:   File "/root/anaconda3/envs/env_blt/lib/python3.10/site-packages/torch/_inductor/codecache.py", line 2876, in load_by_key_path
[rank0]:     mod = _reload_python_module(key, path)
[rank0]:   File "/root/anaconda3/envs/env_blt/lib/python3.10/site-packages/torch/_inductor/runtime/compile_tasks.py", line 45, in _reload_python_module
[rank0]:     exec(code, mod.__dict__, mod.__dict__)
[rank0]:   File "/tmp/torchinductor_root/bq/cbqzzig23hs7zhenym2l2nnqv4qehtyehoqv7y4xlhjd22xkzk3b.py", line 474, in <module>
[rank0]:     async_compile.wait(globals())
[rank0]:   File "/root/anaconda3/envs/env_blt/lib/python3.10/site-packages/torch/_inductor/async_compile.py", line 276, in wait
[rank0]:     scope[key] = result.result()
[rank0]:   File "/root/anaconda3/envs/env_blt/lib/python3.10/site-packages/torch/_inductor/codecache.py", line 3341, in result
[rank0]:     result = self.future.result()
[rank0]:   File "/root/anaconda3/envs/env_blt/lib/python3.10/concurrent/futures/_base.py", line 458, in result
[rank0]:     return self.__get_result()
[rank0]:   File "/root/anaconda3/envs/env_blt/lib/python3.10/concurrent/futures/_base.py", line 403, in __get_result
[rank0]:     raise self._exception
[rank0]: torch._inductor.compile_worker.subproc_pool.SubprocException: An exception occurred in a subprocess:

[rank0]: triton.compiler.errors.CompilationError: at 56:48:
[rank0]:         m = offs_m
[rank0]:         n = offs_n

[rank0]:     post_mod_scores = (qk)


[rank0]:     if CHECK_BLOCK_BOUNDARY:
[rank0]:         # Mask out the elements that are out of the KV_LEN for non divisible seqlen.
[rank0]:         post_mod_scores = tl.where(offs_n < KV_LEN, post_mod_scores, float("-inf"))

[rank0]:     if not IS_FULL_BLOCKS:
[rank0]:         mask_mod_output = tl.load(in_ptr8 + (m)*s10 + (n))
[rank0]:                                                 ^
[rank0]: NameError('s10 is not defined')

[rank0]: The above exception was the direct cause of the following exception:

[rank0]: triton.compiler.errors.CompilationError: at 57:28:
[rank0]:                 acc, l_i, m_i,
[rank0]:                 # Offsets
[rank0]:                 off_z, off_h, offs_m, offs_n,
[rank0]:                 MATMUL_PRECISION, RCP_LN2,
[rank0]:                 IS_FULL_BLOCKS,
[rank0]:             )
[rank0]:         else:
[rank0]:             # Benchmark shows even we applied mod & mask to each block for non divisible seqlen,
[rank0]:             # it's on par or slightly faster than only applying to the last block in fwd.
[rank0]:             # However, we choose different strategy for bwd, where we only apply mod & mask
[rank0]:             # to the last block because it's faster a lot.
[rank0]:             acc, l_i, m_i = forward_block_mn(
[rank0]:                             ^

[rank0]: The above exception was the direct cause of the following exception:

[rank0]: Traceback (most recent call last):
[rank0]:   File "/root/anaconda3/envs/env_blt/lib/python3.10/site-packages/torch/_inductor/compile_worker/subproc_pool.py", line 270, in do_job
[rank0]:     result = job()
[rank0]:   File "/root/anaconda3/envs/env_blt/lib/python3.10/site-packages/torch/_inductor/runtime/compile_tasks.py", line 68, in _worker_compile_triton
[rank0]:     load_kernel().precompile(warm_cache_only=True)
[rank0]:   File "/root/anaconda3/envs/env_blt/lib/python3.10/site-packages/torch/_inductor/runtime/triton_heuristics.py", line 244, in precompile
[rank0]:     compiled_binary, launcher = self._precompile_config(
[rank0]:   File "/root/anaconda3/envs/env_blt/lib/python3.10/site-packages/torch/_inductor/runtime/triton_heuristics.py", line 428, in _precompile_config
[rank0]:     triton.compile(*compile_args, **compile_kwargs),
[rank0]:   File "/root/anaconda3/envs/env_blt/lib/python3.10/site-packages/triton/compiler/compiler.py", line 276, in compile
[rank0]:     module = src.make_ir(options, codegen_fns, context)
[rank0]:   File "/root/anaconda3/envs/env_blt/lib/python3.10/site-packages/triton/compiler/compiler.py", line 113, in make_ir
[rank0]:     return ast_to_ttir(self.fn, self, context=context, options=options, codegen_fns=codegen_fns)
[rank0]: triton.compiler.errors.CompilationError: at 148:20:
[rank0]:     )
[rank0]:     V_block_ptr = tl.make_block_ptr(
[rank0]:         base=V,
[rank0]:         shape=(KV_LEN, V_HEAD_DIM),
[rank0]:         strides=(stride_vn, stride_vk),
[rank0]:         offsets=(kv_start, 0),
[rank0]:         block_shape=(BLOCK_N, V_HEAD_DIM),
[rank0]:         order=(1, 0)
[rank0]:     )
[rank0]:     offs_n = kv_start + tl.arange(0, BLOCK_N)

[rank0]:     acc, l_i, m_i = forward_inner(
[rank0]:                     ^


[rank0]: The above exception was the direct cause of the following exception:

[rank0]: Traceback (most recent call last):
[rank0]:   File "/root/anaconda3/envs/env_blt/lib/python3.10/runpy.py", line 196, in _run_module_as_main
[rank0]:     return _run_code(code, main_globals, None,
[rank0]:   File "/root/anaconda3/envs/env_blt/lib/python3.10/runpy.py", line 86, in _run_code
[rank0]:     exec(code, run_globals)
[rank0]:   File "/root/Desktop/Project/blt-main/bytelatent/train.py", line 812, in <module>
[rank0]:     main()
[rank0]:   File "/root/Desktop/Project/blt-main/bytelatent/train.py", line 808, in main
[rank0]:     train(train_args)
[rank0]:   File "/root/Desktop/Project/blt-main/bytelatent/train.py", line 475, in train
[rank0]:     pred = model(
[rank0]:   File "/root/anaconda3/envs/env_blt/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1736, in _wrapped_call_impl
[rank0]:     return self._call_impl(*args, **kwargs)
[rank0]:   File "/root/anaconda3/envs/env_blt/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1844, in _call_impl
[rank0]:     return inner()
[rank0]:   File "/root/anaconda3/envs/env_blt/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1790, in inner
[rank0]:     result = forward_call(*args, **kwargs)
[rank0]:   File "/root/Desktop/Project/blt-main/bytelatent/model/blt.py", line 1023, in forward
[rank0]:     output, _ = self.local_decoder(
[rank0]:   File "/root/anaconda3/envs/env_blt/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1736, in _wrapped_call_impl
[rank0]:     return self._call_impl(*args, **kwargs)
[rank0]:   File "/root/anaconda3/envs/env_blt/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1747, in _call_impl
[rank0]:     return forward_call(*args, **kwargs)
[rank0]:   File "/root/Desktop/Project/blt-main/bytelatent/model/local_models.py", line 390, in forward
[rank0]:     h_cross = self.cross_attn_layers[i](
[rank0]:   File "/root/anaconda3/envs/env_blt/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1736, in _wrapped_call_impl
[rank0]:     return self._call_impl(*args, **kwargs)
[rank0]:   File "/root/anaconda3/envs/env_blt/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1844, in _call_impl
[rank0]:     return inner()
[rank0]:   File "/root/anaconda3/envs/env_blt/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1790, in inner
[rank0]:     result = forward_call(*args, **kwargs)
[rank0]:   File "/root/Desktop/Project/blt-main/bytelatent/model/latent_transformer.py", line 105, in forward
[rank0]:     output = flex_attention_comp(xq, xk, xv, block_mask=mask)
[rank0]:   File "/root/anaconda3/envs/env_blt/lib/python3.10/site-packages/torch/_dynamo/eval_frame.py", line 465, in _fn
[rank0]:     return fn(*args, **kwargs)
[rank0]:   File "/root/anaconda3/envs/env_blt/lib/python3.10/site-packages/torch/_dynamo/convert_frame.py", line 1269, in __call__
[rank0]:     return self._torchdynamo_orig_callable(
[rank0]:   File "/root/anaconda3/envs/env_blt/lib/python3.10/site-packages/torch/_dynamo/convert_frame.py", line 1064, in __call__
[rank0]:     result = self._inner_convert(
[rank0]:   File "/root/anaconda3/envs/env_blt/lib/python3.10/site-packages/torch/_dynamo/convert_frame.py", line 526, in __call__
[rank0]:     return _compile(
[rank0]:   File "/root/anaconda3/envs/env_blt/lib/python3.10/site-packages/torch/_dynamo/convert_frame.py", line 924, in _compile
[rank0]:     guarded_code = compile_inner(code, one_graph, hooks, transform)
[rank0]:   File "/root/anaconda3/envs/env_blt/lib/python3.10/site-packages/torch/_dynamo/convert_frame.py", line 666, in compile_inner
[rank0]:     return _compile_inner(code, one_graph, hooks, transform)
[rank0]:   File "/root/anaconda3/envs/env_blt/lib/python3.10/site-packages/torch/_utils_internal.py", line 87, in wrapper_function
[rank0]:     return function(*args, **kwargs)
[rank0]:   File "/root/anaconda3/envs/env_blt/lib/python3.10/site-packages/torch/_dynamo/convert_frame.py", line 699, in _compile_inner
[rank0]:     out_code = transform_code_object(code, transform)
[rank0]:   File "/root/anaconda3/envs/env_blt/lib/python3.10/site-packages/torch/_dynamo/bytecode_transformation.py", line 1322, in transform_code_object
[rank0]:     transformations(instructions, code_options)
[rank0]:   File "/root/anaconda3/envs/env_blt/lib/python3.10/site-packages/torch/_dynamo/convert_frame.py", line 219, in _fn
[rank0]:     return fn(*args, **kwargs)
[rank0]:   File "/root/anaconda3/envs/env_blt/lib/python3.10/site-packages/torch/_dynamo/convert_frame.py", line 634, in transform
[rank0]:     tracer.run()
[rank0]:   File "/root/anaconda3/envs/env_blt/lib/python3.10/site-packages/torch/_dynamo/symbolic_convert.py", line 2796, in run
[rank0]:     super().run()
[rank0]:   File "/root/anaconda3/envs/env_blt/lib/python3.10/site-packages/torch/_dynamo/symbolic_convert.py", line 983, in run
[rank0]:     while self.step():
[rank0]:   File "/root/anaconda3/envs/env_blt/lib/python3.10/site-packages/torch/_dynamo/symbolic_convert.py", line 895, in step
[rank0]:     self.dispatch_table[inst.opcode](self, inst)
[rank0]:   File "/root/anaconda3/envs/env_blt/lib/python3.10/site-packages/torch/_dynamo/symbolic_convert.py", line 2987, in RETURN_VALUE
[rank0]:     self._return(inst)
[rank0]:   File "/root/anaconda3/envs/env_blt/lib/python3.10/site-packages/torch/_dynamo/symbolic_convert.py", line 2972, in _return
[rank0]:     self.output.compile_subgraph(
[rank0]:   File "/root/anaconda3/envs/env_blt/lib/python3.10/site-packages/torch/_dynamo/output_graph.py", line 1117, in compile_subgraph
[rank0]:     self.compile_and_call_fx_graph(tx, list(reversed(stack_values)), root)
[rank0]:   File "/root/anaconda3/envs/env_blt/lib/python3.10/site-packages/torch/_dynamo/output_graph.py", line 1369, in compile_and_call_fx_graph
[rank0]:     compiled_fn = self.call_user_compiler(gm)
[rank0]:   File "/root/anaconda3/envs/env_blt/lib/python3.10/site-packages/torch/_dynamo/output_graph.py", line 1416, in call_user_compiler
[rank0]:     return self._call_user_compiler(gm)
[rank0]:   File "/root/anaconda3/envs/env_blt/lib/python3.10/site-packages/torch/_dynamo/output_graph.py", line 1465, in _call_user_compiler
[rank0]:     raise BackendCompilerFailed(self.compiler_fn, e) from e
[rank0]: torch._dynamo.exc.BackendCompilerFailed: backend='inductor' raised:
[rank0]: SubprocException: An exception occurred in a subprocess:

[rank0]: triton.compiler.errors.CompilationError: at 56:48:
[rank0]:         m = offs_m
[rank0]:         n = offs_n

[rank0]:     post_mod_scores = (qk)


[rank0]:     if CHECK_BLOCK_BOUNDARY:
[rank0]:         # Mask out the elements that are out of the KV_LEN for non divisible seqlen.
[rank0]:         post_mod_scores = tl.where(offs_n < KV_LEN, post_mod_scores, float("-inf"))

[rank0]:     if not IS_FULL_BLOCKS:
[rank0]:         mask_mod_output = tl.load(in_ptr8 + (m)*s10 + (n))
[rank0]:                                                 ^
[rank0]: NameError('s10 is not defined')

[rank0]: The above exception was the direct cause of the following exception:

[rank0]: triton.compiler.errors.CompilationError: at 57:28:
[rank0]:                 acc, l_i, m_i,
[rank0]:                 # Offsets
[rank0]:                 off_z, off_h, offs_m, offs_n,
[rank0]:                 MATMUL_PRECISION, RCP_LN2,
[rank0]:                 IS_FULL_BLOCKS,
[rank0]:             )
[rank0]:         else:
[rank0]:             # Benchmark shows even we applied mod & mask to each block for non divisible seqlen,
[rank0]:             # it's on par or slightly faster than only applying to the last block in fwd.
[rank0]:             # However, we choose different strategy for bwd, where we only apply mod & mask
[rank0]:             # to the last block because it's faster a lot.
[rank0]:             acc, l_i, m_i = forward_block_mn(
[rank0]:                             ^

[rank0]: The above exception was the direct cause of the following exception:

[rank0]: Traceback (most recent call last):
[rank0]:   File "/root/anaconda3/envs/env_blt/lib/python3.10/site-packages/torch/_inductor/compile_worker/subproc_pool.py", line 270, in do_job
[rank0]:     result = job()
[rank0]:   File "/root/anaconda3/envs/env_blt/lib/python3.10/site-packages/torch/_inductor/runtime/compile_tasks.py", line 68, in _worker_compile_triton
[rank0]:     load_kernel().precompile(warm_cache_only=True)
[rank0]:   File "/root/anaconda3/envs/env_blt/lib/python3.10/site-packages/torch/_inductor/runtime/triton_heuristics.py", line 244, in precompile
[rank0]:     compiled_binary, launcher = self._precompile_config(
[rank0]:   File "/root/anaconda3/envs/env_blt/lib/python3.10/site-packages/torch/_inductor/runtime/triton_heuristics.py", line 428, in _precompile_config
[rank0]:     triton.compile(*compile_args, **compile_kwargs),
[rank0]:   File "/root/anaconda3/envs/env_blt/lib/python3.10/site-packages/triton/compiler/compiler.py", line 276, in compile
[rank0]:     module = src.make_ir(options, codegen_fns, context)
[rank0]:   File "/root/anaconda3/envs/env_blt/lib/python3.10/site-packages/triton/compiler/compiler.py", line 113, in make_ir
[rank0]:     return ast_to_ttir(self.fn, self, context=context, options=options, codegen_fns=codegen_fns)
[rank0]: triton.compiler.errors.CompilationError: at 148:20:
[rank0]:     )
[rank0]:     V_block_ptr = tl.make_block_ptr(
[rank0]:         base=V,
[rank0]:         shape=(KV_LEN, V_HEAD_DIM),
[rank0]:         strides=(stride_vn, stride_vk),
[rank0]:         offsets=(kv_start, 0),
[rank0]:         block_shape=(BLOCK_N, V_HEAD_DIM),
[rank0]:         order=(1, 0)
[rank0]:     )
[rank0]:     offs_n = kv_start + tl.arange(0, BLOCK_N)

[rank0]:     acc, l_i, m_i = forward_inner(
[rank0]:                     ^


[rank0]: Set TORCH_LOGS="+dynamo" and TORCHDYNAMO_VERBOSE=1 for more information


[rank0]: You can suppress this exception and fall back to eager by setting:
[rank0]:     import torch._dynamo
[rank0]:     torch._dynamo.config.suppress_errors = True

Process blt_data_loader:
Traceback (most recent call last):
  File "/root/anaconda3/envs/env_blt/lib/python3.10/multiprocessing/process.py", line 314, in _bootstrap
    self.run()
  File "/root/anaconda3/envs/env_blt/lib/python3.10/multiprocessing/process.py", line 108, in run
    self._target(*self._args, **self._kwargs)
  File "/root/Desktop/Project/blt-main/bytelatent/data/iterators/multiprocess_iterator.py", line 84, in start_work_from_state
    batch_queue.put(item, timeout=0.1)
  File "<string>", line 2, in put
  File "/root/anaconda3/envs/env_blt/lib/python3.10/multiprocessing/managers.py", line 818, in _callmethod
    kind, result = conn.recv()
  File "/root/anaconda3/envs/env_blt/lib/python3.10/multiprocessing/connection.py", line 250, in recv
    buf = self._recv_bytes()
  File "/root/anaconda3/envs/env_blt/lib/python3.10/multiprocessing/connection.py", line 414, in _recv_bytes
    buf = self._recv(4)
  File "/root/anaconda3/envs/env_blt/lib/python3.10/multiprocessing/connection.py", line 383, in _recv
    raise EOFError
EOFError

**Comment by EntilZha** [comment]
Arrow iterator can take json files now and the attention issue is discussed here https://github.com/facebookresearch/blt/issues/6#issuecomment-2727890119

If there are additional issues with train.py and arrow/json files, please open a new issue to keep discussion more narrow

**Comment by zbdx-lcy** [comment]
> 作者您好！在您的指导下，培训过程进展顺利。但是，我已经训练了熵模型，目前正在运行 preprocess_entropies.py 脚本。我有一个问题：用于训练熵模型的数据集和运行 preprocess_entropies.py 脚本时用于生成 Arrow 文件的数据集有什么区别？是：两个进程需要相同的数据集吗？用于这两个过程的数据集完全不同？用于训练 BLT 模型的数据集包括用于训练熵模型的数据集？我目前对此有点不清楚，非常感谢您的帮助。非常感谢！

Hello, have you solved this problem? I also have the exact same question to ask.
